Refectoring that aims to fix Rekordbox crash on macOS #10608

daschuer opened the pull request:
We have a confirmed crasher on macOS regarding Reckordbox #10608. 
Because it is not reproducible a did a tight review of the memory management in the surrounding code. 

I have discovered a suspicious pointer handling that has been introduced because of the dynamic nature of the tree view in case of the Reckordbox feature. Unprotected raw pointers are carried by the QModelIndex and may become dangling after a tree rebuild. I have introduces smart pointers as best as possible without a complete rewrite and deleted the possible dangling pointers in 3260e9c914fa34b5aa8fd5ae795084151ffe708a and 83973e1a80c9b0cb6c921b3427987778ff6ef4ed

In addition I have cleand up the code around the search columns that has suffered some rotting.  

Even though I have still no evidence that this PR fixes any crash there is a chance. If not this is IMHO a useful refactoring. 
That's why I have target this to 2.3. If we think it is to big for a bug fix we can merge it to 2.4.

daschuer committed:
Use QStringLiteral for Library columns

daschuer committed:
Clear LastRightClickIndex when removing tree items.

It can become a dangling pointer

daschuer committed:
Use std::unique_ptr when creating the library tree

daschuer committed:
Reset m_pLastRightClickedItem before removing treeItems.

This way it cannot become a dangling pointer

daschuer committed:
Make use of unique_ptr and std::move

This avoids memory leaks

daschuer committed:
Inmprove initialization of library columns

daschuer committed:
Use std::move in BaseSqlTableModel::setTable

daschuer committed:
Move BaseTrackCache::setSearchColumns to constructor, to create the object at once.

daschuer committed:
Clean up search columns of the Serato and Rekordbox feature to match the behavior of the native Mixxx library

daschuer committed:
remove unused m_searchColumnIndices

daschuer committed:
Remove unused include

daschuer committed:
Browsetablemodel: Simplify adding search columns. Not yet enabled.

daschuer committed:
make future result const before iteration

Swiftb0y commented:
I took a quick look at this PR and I agree that this likely won't really fix any crash but it highlights the unsafe pointer handling by use of unsafe unique_ptr APIs. I will conduct a full review asap.

Swiftb0y commented on src/library/baseplaylistfeature.cpp:
```diff
@@ -633,17 +633,17 @@ QModelIndex BasePlaylistFeature::constructChildModel(int selected_id) {
 
-        decorateChild(item, playlistId);
-        data_list.append(item);
+        decorateChild(pItem.get(), playlistId);
```
`decorateChild` seems to be trivial to refactor to use a `gsl::not_null<TreeItem>`. That would help clarify that its non-owning (at least I think we established that `gsl::not_null` pointers are non-owning, didn't we?)

daschuer replied:
gsl is unfortunately not available in the 2.3 branch.

Swiftb0y replied:
Mhmm, then it would be nice if you could document the non-owning nature of pointer in `BasePlaylistFeature`

daschuer replied:
Done.

Swiftb0y commented on src/library/baseexternallibraryfeature.h:
```diff
@@ -26,9 +27,10 @@ class BaseExternalLibraryFeature : public LibraryFeature {
   protected:
     // Must be implemented by external Libraries copied to Mixxx DB
-    virtual BaseSqlTableModel* getPlaylistModelForPlaylist(const QString& playlist) {
+    virtual std::unique_ptr<BaseSqlTableModel> createPlaylistModelForPlaylist(
+            const QString& playlist) {
```
prefer `QStringView` over `const QString&` afaik. 
```suggestion
    virtual std::unique_ptr<BaseSqlTableModel> createPlaylistModelForPlaylist(
            QStringView playlist) {
```

daschuer replied:
I have just tried that, but it fails, because of the final `QVariant(const QString&)` call. 
This requires a QStringView::toString() which performs a deep copy.

Swiftb0y replied:
Mhmm shame, I guess replacing that with a `QVariant(QStringView)` is not possible?

Swiftb0y commented on src/library/rekordbox/rekordboxfeature.cpp:
```diff
@@ -162,6 +162,8 @@ bool dropTable(QSqlDatabase& database, const QString& tableName) {
 
 // This function is executed in a separate thread other than the main thread
+// The returned list owns the pointers, but we can't use a unique_ptr because
+// the result is passed by a const reference
```
passed where? I don't really understand the problem or how unique_ptr interferes.

daschuer replied:
.. when passed inside QFuture;

daschuer replied:
I will improve the comments

Swiftb0y commented on src/library/serato/seratofeature.cpp:
```diff
@@ -1059,9 +1062,15 @@ void SeratoFeature::activateChild(const QModelIndex& index) {
 void SeratoFeature::onSeratoDatabasesFound() {
-    QList<TreeItem*> foundDatabases = m_databasesFuture.result();
-    TreeItem* root = m_childModel.getRootItem();
+    std::vector<std::unique_ptr<TreeItem>> foundDatabases;
+    const QList<TreeItem*> result = m_databasesFuture.result();
+    for (const auto& pDatabaseFound : result) {
+        foundDatabases.emplace_back(pDatabaseFound);
+    }
```
Does iterator-based construction also work?
```suggestion
    const QList<TreeItem*> result = m_databasesFuture.result();
    std::vector<std::unique_ptr<TreeItem>> foundDatabases(result.begin(), result.end());
```

Swiftb0y commented on src/library/treeitem.cpp:
```diff
@@ -82,34 +82,22 @@ void TreeItem::initFeatureRecursively(LibraryFeature* pFeature) {
         QString label,
         QVariant data) {
-    auto pNewChild = std::make_unique<TreeItem>(
+    std::unique_ptr<TreeItem> pNewChild = std::make_unique<TreeItem>(
```
this feels redundant. I thought declaration with `auto` was fine since the return type is easily inferred from `make_unique`.

daschuer replied:
Done.

daschuer committed:
Remove pointless comments

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>

daschuer committed:
Make use of SharedPointer::create to avoid a literally new,

daschuer commented on src/library/basesqltablemodel.cpp:
```diff
@@ -334,16 +334,16 @@ void BaseSqlTableModel::select() {
+void BaseSqlTableModel::setTable(QString tableName,
+        QString idColumn,
+        QStringList tableColumns,
         QSharedPointer<BaseTrackCache> trackSource) {
     if (sDebug) {
         qDebug() << this << "setTable" << tableName << tableColumns << idColumn;
     }
-    m_tableName = tableName;
-    m_idColumn = idColumn;
-    m_tableColumns = tableColumns;
+    m_tableName = std::move(tableName);
+    m_idColumn = std::move(idColumn);
+    m_tableColumns = std::move(tableColumns);
 
```
Your https://godbolt.org/z/TEvY5e7ff does not give the correct impression, because the structure is empty and the compile cannot optimize the functions away because of the message output and the disabled optimization.
I have played with a more realistic https://godbolt.org/z/sdGz5nYbW
Here is turns out that the copy and rvalue function have the same assembler code, while the original const ref function creates a deep copy. With the copy version, the caller is free to copy or move the value in. That's why I prefer the copy version.  

Finally it depends also weather the object is cheap to move or not. In case of QT object with the Pimpl pattern moving is just passing a pointer. So even if there is an extra move for some reasons, it does not matter but keeps the interface simple.

daschuer replied:
I did a second look why the value and the lref version are equal. It is bacuse copying an lref is also copying a pointer. 
So if moving is also only a matter of a pointer copy from an inlined move constructor, it is the same. 

In this case here the lref version requires a copy for some library l features in the calling code. To avoid that, we would need a const ref overload as well. That's why the copy version is more universal with the cost of an extra move in cases where it cannot inlined.

Swiftb0y replied:
> Your https://godbolt.org/z/TEvY5e7ff does not give the correct impression, because the structure is empty and the compile cannot optimize the functions away because of the message output and the disabled optimization.

Not really, the fact that there are user defined copy-/move-constructors at all inhibited optimization (since the type is no longer trivially move-/copyable). Looking at the assembly all of the calls still get inlined. Regardless, that was not the point I was trying to make. I was rather talking about the semantics present in C++ which get enforced by the language. 

>  I have played with a more realistic https://godbolt.org/z/sdGz5nYbW

In the case of Qt-classes this example is even more unrealistic IMO because it relies on the optimizer. For the optimizer to do its job it relies on being able to see the source, which is only the case for move assignment and the move constructor for many Qt classes (see the [move assign and move ctor in Qstring header](https://codebrowser.dev/qt5/qtbase/src/corelib/text/qstring.h.html#_ZN7QStringC1EOS_). 

Anyways, I still don't understand why one should not take by `&&` when there is an unconditional move in the body. If the caller really doesn't want the value to be moved out of, they can explicitly copy it into a temporary later. That way you only get a single move by default and if you really need the copy, the caller can create it explicitly beforehand.

daschuer replied:
> I was rather talking about the semantics present in C++ which get enforced by the language.

From that point of view your conclusion is correct.

> Anyway, I still don't understand why one should not take by && when there is an unconditional move in the body.

For me it is a good rule of thumb to use copy by value in case the function keeps the value and use const ref otherwise. This allows all things like: 
```
A a;
function(a)
function(move(a))
function(otherFunctionReturingA())
```
Which results in a bare minimum of deep copies and works with one function signature. 
This works of cause only optimal if `A::A(A&& a)` is equal cheap compared to copy an address.     
Otherwise you need to take care and implement the `const ref` version and the `non const lref` overload like you have proposed as universal solution.

Swiftb0y replied:
I disagree with your conclusion. The proposed solution of taking `A&&` does not need any overloads, uses the same amount of copies and even fewer moves. The only difference is that the caller then explicitly needs to specify whether to copy or to move (which I actually prefer over the default value semantics). This demonstrates the phenomenon:
https://godbolt.org/z/YEexffYbn

Swiftb0y replied:
In details for the const ref. If you want to call a rvalue-only function with a const ref, you need to copy explicitly (good imo) while if you do the same for a function taking by value, the copy is implicit. If you do the same for a non-const lref, the compiler still complains, forcing you to either make an explicit copy or explicit move. required implicit moves are still performed while implicit copies are eliminated.

daschuer replied:
True, however that would be an unusual style.

Swiftb0y replied:
Fair. Though I'd be very keen to know why since I don't see any downsights to this style.

daschuer replied:
This is just because of legacy.

Swiftb0y replied:
Can you elaborate? What's there to say against using the rvalue-ref-only solution?

daschuer replied:
rvalue references is a "new" c++11 feature, and not a first class citizen like in Rust. When using them from a lvalue, `std::move(lvalue)` is mandatory. Requiring a mandatory `T(lvalue)` for a function would be a new concept, exposing an unnecessarily and implementation detail. Normally you see a `function(A a)` with a `function(const A& a)` to hide that from the calling code. Just providing `function(A&& a)` is breaking the normal `function(a)` call. The user is forced to consider if he wants to keep that value or throw it away. This decision should be optional only for those who know what they are doing.

We had a similar discussion some time ago that leads to this rule: 
https://github.com/mixxxdj/mixxx/wiki/Coding-Guidelines#r-value-references--move-constructors--move-assignment-operators

Some details and reasoning can be found here:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#fcall-parameter-passing

Swiftb0y replied:
Mhmm, thank you for these resources. I agree that having to specify whether to move or copy is a breaking change and also not super easy to understand with the usual obtuse error messages generated by c++ compilers. That said, for this case, even the core guidelines recommend taking by it by `&&` universally for this particular case where we just unconditionally move. You can see that in the "Advanced Parameter passing" figure in the "in + move" case. This is explained in more details in [F.18  For “will-move-from” parameters, pass by X&& and std::move the parameter](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f18-for-will-move-from-parameters-pass-by-x-and-stdmove-the-parameter). There is an exception for "Unique owner types that are move-only and cheap-to-move, such as unique_ptr" but I don't think Qt's implicitly-shared classes really qualify for that since they're not types that are only concerned with ownership.

daschuer replied:
Taking a && is an advanced technik where the guide is clear about: 

> Use the advanced techniques only after demonstrating need, and document that need in a comment.

We require the same in the Mixxxx code guiline.

Swiftb0y replied:
I acknowledge that. I'm fine with that as long as the Type in question is cheap to move (which includes the implicitly-shared Qt containers). Is that ok with you?

daschuer replied:
Yes. Maybe after that long discussion we need to update the single sentences in the Mixxx guideline.

Holzhaus commented on src/library/banshee/bansheeplaylistmodel.cpp:
```diff
@@ -34,11 +34,34 @@
+const QString kTrackId = QStringLiteral(CLM_TRACK_ID);
+const QString kViewOrder = QStringLiteral(CLM_VIEW_ORDER);
+const QString KArtist = QStringLiteral(CLM_ARTIST);
+const QString KTitel = QStringLiteral(CLM_TITLE);
```
```suggestion
const QString KTitle = QStringLiteral(CLM_TITLE);
```

Swiftb0y replied:
capitalization
```suggestion
const QString kArtist = QStringLiteral(CLM_ARTIST);
const QString kTitel = QStringLiteral(CLM_TITLE);
```

daschuer committed:
Fix naming of contsnats

daschuer committed:
Prefere QModelIndex(); when resetting stored indexes.

daschuer committed:
use std::make_unique when initalize m_pQueryParser

daschuer committed:
Improve the  comment about QFuture

daschuer committed:
Use iterator-based construction

daschuer committed:
Remove redundant type name

daschuer commented on src/library/treeitem.cpp:
```diff
@@ -82,34 +82,22 @@ void TreeItem::initFeatureRecursively(LibraryFeature* pFeature) {
-    DEBUG_ASSERT(row >= 0);
-    DEBUG_ASSERT(row < m_children.size());
-    delete m_children.takeAt(row);
+    TreeItem* pRet = pNewChild.get();
+    insertChild(m_children.size(), std::move(pNewChild));
+    return pRet;
```
I have tried all sorts of solution but finally gave up. The issue is that QModelIndex takes a raw pointer internally nad you cannot convert a raw pointer back to a shared pointer. 
A entire smart pointer based solution requires a big refactoring that is out of scope of this PR.

Swiftb0y replied:
What QModelIndex do you mean exactly?

daschuer replied:
Here: 
https://github.com/mixxxdj/mixxx/blob/eff27abd571ac3e1b273ad4aef6407eceee0cfee/src/library/autodj/autodjfeature.cpp#L292

Swiftb0y replied:
Mhmm ok, i see the problem. Can you add a comment that says the pointer that is being returned is non-owning and only valid for the lifetime duration of the `TreeItem`?

Swiftb0y replied:
I'm not quite sure if that is completely correct since the pointer would also be invalid if children get removed...

daschuer replied:
I will add a comment. When the child is removed, the TreeItem is deleted and this pointer becomes dangling ...
dangerous .... This could be IMHO the source of the crash. In this PR I took care to delete those dangling pointers, whenever the tree is touched.

daschuer commented on src/library/treeitem.cpp:
```diff
@@ -82,34 +82,22 @@ void TreeItem::initFeatureRecursively(LibraryFeature* pFeature) {
 }
 
-void TreeItem::insertChildren(int row, QList<TreeItem*>& children) {
+void TreeItem::insertChildren(int row, std::vector<std::unique_ptr<TreeItem>>&& children) {
```
This is a 2.3 PR where no std::span is available. I also think it is the best solution to just pass the object we have instead of dealing with another type.

Swiftb0y replied:
Mhmm, the absence of span is not great. I guess thats why refactorings don't really belong into the stable branch.
I disagree with the statement that dealing with another type is a bad idea, IMO in this case it would be an improvement because it improves the communication of the intend and also decreases coupling. But since `std::span` is not available (and using iterators is somewhat complicated) I'm fine with keeping it as is. A TODO comment would be nice though.

daschuer replied:
std::span it is a pointer and a size. It has been introduced to manage plain arrays in a safer way. Why would it be better to access a std::vector a like this. The vector has that already. If you like to express that it is borrowed, you can use a const refrence. 

std::span is usefull when you are dealing with only some elements of the vector. But it is useless for all other containers where we have not a consecutive memory layout.

In this case I like that the memory for the unique_ptr itself is also passed to the consuming function. This way you can easy see that the vector with its elements is destroyed.

Swiftb0y replied:
> std::span it is a pointer and a size. It has been introduced to manage plain arrays in a safer way. Why would it be better to access a std::vector a like this.

Well std::span is a generic view over any container that stores its element consecutively in memory. So its generic over a vector and over an array (and possibly QVarLengthArray and many other types that meet that criteria). 

> If you like to express that it is borrowed, you can use a const refrence.

Well not really, I want to express that I want to mutably access the container without growing its size. For that, `const std::span<T>` (note the non-const T) is best. 

> std::span is useful when you are dealing with only some elements of the vector. But it is useless for all other containers where we have not a consecutive memory layout.

Some elements is still just as useful as all elements. Yes if we had a container with a non-consequitive memory layout, then iterators (or ranges) would be best choice, but as I said, that comes with some complexity which you might want to avoid. 

> In this case I like that the memory for the unique_ptr itself is also passed to the consuming function. This way you can easy see that the vector with its elements is destroyed.

Yes thats a fair point, I just don't like the vector because it forces the caller to allocate a vector even though we don't actually care about the fact that its a vector, we just care about the unique_ptr's within it.

daschuer replied:
> Well not really, I want to express that I want to mutably access the container without growing its size. For that, const std::span<T> (note the non-const T) is best.

Ok, that makes sense. I knowledge also that this simplifies the single row case as you pointed out.

Swiftb0y replied:
Thank you. As you pointed out, there is no std::span or gsl::span in 2.3 so can you add a TODO to the function signatures of `TreeItem::insertChildren` and `TreeItemModel::insertTreeItemRows` that reminds us to refactor that in 2.4? Then I'll consider this resolved.

daschuer commented on src/library/treeitemmodel.cpp:
```diff
@@ -164,22 +164,21 @@ TreeItem* TreeItemModel::setRootItem(std::unique_ptr<TreeItem> pRootItem) {
+ * make sure you have initialized.
  */
 void TreeItemModel::insertTreeItemRows(
-        QList<TreeItem*>& rows,
+        std::vector<std::unique_ptr<TreeItem>>&& rows,
```
the heap allocations happens in the calling code. For instance in BasePlaylistFeature::constructChildModel Line 640

Swiftb0y replied:
Yes, but there are at least two locations where it could be avoided if we instead took a span instead of a rvalue-ref to a vector because the caller could just use an array and create a span from that.

https://github.com/mixxxdj/mixxx/pull/11230/files#diff-0e02fa72cc2bf68948f489976e3edf7634fa84b5f03b77e545593018565b81dbR218-R222

https://github.com/mixxxdj/mixxx/pull/11230/files#diff-16fa56a50091f69845e3552ebd062b6902ffe9ed849825809bab4b75106acc89R154-R156

daschuer commented:
I think I have addressed all comments. @Swiftb0y some of your ideas need to go to 2.4. 
Do you have an opinion for the target branch of this PR? At least https://github.com/mixxxdj/mixxx/commit/3260e9c914fa34b5aa8fd5ae795084151ffe708a and https://github.com/mixxxdj/mixxx/commit/83973e1a80c9b0cb6c921b3427987778ff6ef4ed should go to 2.3.

Swiftb0y commented on src/library/rekordbox/rekordboxfeature.cpp:
```diff
@@ -163,7 +163,7 @@ bool dropTable(QSqlDatabase& database, const QString& tableName) {
 // This function is executed in a separate thread other than the main thread
 // The returned list owns the pointers, but we can't use a unique_ptr because
-// the result is passed by a const reference
+// the result is passed by a const reference inside QFuture.
```
thank you for updating the comment. It still didn't make sense for me until I dug a little deeper. As far as I can tell the actual problem is that QFuture needs to be parameterized on a type that has copy constructors? If so, that also highlights a potential safety issue. If the QList is copyable and it just copies the contained pointers, that would lead to dangling pointers with double free's etc. Right?

daschuer replied:
>  QFuture needs to be parameterized on a type that has copy constructors

Yes. 

> If so, that also highlights a potential safety issue. If the QList is copyable and it just copies the contained pointers, that would lead to dangling pointers with double free's etc. Right?

Dangling pointers are freed pointers, that are not null. This does not happens in this case, because the QList only owns the pointers not the object. When the copy in the worker thread is deleted only the memory location of the pointers is deleted, not the objects the have been pointing to. They are accessed in the main thread and made owning and managed by the `std::vector<std::unique_ptr>>` before second copy of the original QList is deleted.

daschuer replied:
Comment updated.

Swiftb0y replied:
> When the copy in the worker thread is deleted only the memory location of the pointers is deleted, not the objects the have been pointing to.

Ah, I assumed the QList was deleting the objects pointed to the pointers, that makes sense.

Swiftb0y commented on src/library/serato/seratofeature.cpp:
```diff
@@ -1062,11 +1062,9 @@ void SeratoFeature::activateChild(const QModelIndex& index) {
 
 void SeratoFeature::onSeratoDatabasesFound() {
-    std::vector<std::unique_ptr<TreeItem>> foundDatabases;
+    // std::vector<std::unique_ptr<TreeItem>> foundDatabases;
```
Didn't you mean to remove instead?
```suggestion
```

Swiftb0y reviewed (changes requested):
Thank you, I'll still have to re-review a couple other parts though.

Swiftb0y commented:
> Do you have an opinion for the target branch of this PR? 

2.3 is fine if we leave comments in the code for the stuff that has to be done in 2.4.

daschuer committed:
Remove stary comment

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>

daschuer committed:
Describe the issue with QFuture and std::unique_ptr() a bit more

daschuer committed:
Document decorateChild()

daschuer committed:
use auto in connection with std::make_unique

daschuer commented:
I think I have addressed everything.

Swiftb0y commented:
There are still a couple questions from me and ronso0 which have yet to be answered. I have double checked every comment and IMO everything that is not marked resolved yet is still an unfinished discussion.

daschuer commented on src/library/baseplaylistfeature.cpp:
```diff
@@ -671,6 +671,7 @@ void BasePlaylistFeature::updateChildModel(int playlistId) {
   * Clears the child model dynamically, but the invisible root item remains
   */
 void BasePlaylistFeature::clearChildModel() {
+    m_lastRightClickedIndex = {};
```
Yes, I have mad it more verbose.

daschuer commented:
I think I have added now all missing answers.

Swiftb0y commented:
Thank you, I replied in their respective threads.

daschuer committed:
Add todos for introducing a std::span to simplifx the one element case

daschuer commented:
Done

Swiftb0y commented:
Thank you, there is still another open comment where I requested to document the lifetime of a raw pointer.

daschuer committed:
Add comments about the risk of dangling pointers

daschuer commented:
Done

Swiftb0y commented:
Thank you, can you take care of the failing pre-commit clang-format?

daschuer commented:
Ups, done.

Swiftb0y reviewed (approved):
Thank you, sorry for not merging earlier. Lost track of this PR. Busy week.
